src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py

```diff
@@ -369,29 +392,53 @@ def _purgeTree(self, ctx, baseDirPfn):
         :param ctx:  The gfal2 context object
         :return:     Bool: True if it managed to purge everything, False otherwise
         """
+        # NOTE: It deletes only directories and does not try to unlink any file.
+
         successList = []
 
+        # First test if baseDirPfn is actually a directory entry:
+        try:
+            entryStat = ctx.stat(baseDirPfn)
+        except gfal2.GError as gfalExc:
+            e = sys.exc_info()[1]
+            if e.code == errno.ENOENT:
+                self.logger.error("MISSING baseDir: %s", baseDirPfn)
+                successList.append(False)
+                return all(successList)
+            else:
+                self.logger.error("FAILED baseDir: %s: gfalException: %s", baseDirPfn, str(gfalExc))
+                successList.append(False)
+                return all(successList)
+        if not stat.S_ISDIR(entryStat.st_mode):
+            self.logger.error("The base pfn: %s is not a directory entry.", baseDirPfn)
+            successList.append(False)
+            return all(successList)
+
         if baseDirPfn[-1] != '/':
             baseDirPfn += '/'
 
+        # Second recursively iterate down the tree:
         for dirEntry in ctx.listdir(baseDirPfn):
             if dirEntry in ['.', '..']:
                 continue
             dirEntryPfn = baseDirPfn + dirEntry
             try:
                 entryStat = ctx.stat(dirEntryPfn)
-            except gfal2.GError:
+            except gfal2.GError as gfalExc:
                 e = sys.exc_info()[1]
                 if e.code == errno.ENOENT:
                     self.logger.error("MISSING dirEntry: %s", dirEntryPfn)
                     successList.append(False)
                     return all(successList)
                 else:
```

Comment:
Todor, I see that these blocks are manipulating a file/directory similar to what is done in the method `_purgeTree`. How about we create auxiliary methods for those, such that we can reuse code and even get some simple unit tests done for this?

Reply:
Alan, I do not get your comment here. The line you are commenting on is still inside the `_purgeTree` method. And yes those three blocks they do handle similar errors, but they are doing different operations:
> \# First test if baseDirPfn is actually a directory entry:

which is doing a `ctx.stat()` operation on the top level entry point.   
...

> \# Second recursively iterate down the tree:

which is doing `ctx.listdir()` and the actual recursion
...

> \# Finally remove the baseDir:

which is doing the actual removal `ctx.rmdir()`

Reply:
Sorry @amaltaro I still do not see you point. These 3 blocks are doing 3 different things and are similar (but not completely identical) only on the way how they handle the Gfal exception (under the same circumstances one of them return direct bool and the other two accumulate in a list). Are you suggesting me to create  a function only for handling these exceptions and the pass to it
1. the gfal exception
2. an external list etc 

If so I do not think this would be a good idea that would make us juggling with many conditions ..... 
I did try in the past to go with a single cycle for the whole recursion.... kind of achievable .. but then things get not that  obvious what is happening at the bottom. This separation of three steps seemed the most accurate to me.

Reply:
Todor, you are right. It's not trivial and you are also using the stat response to trigger a recursive call, which I missed when I looked at this code. Sorry for the noise.